<!DOCTYPE html>
<html>
<head>
<title>README.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h2 id="getting-started">Getting Started</h2>
<pre class="hljs"><code><div><span class="hljs-comment"># download compleasm and its dependencies (miniprot and hmmsearch)</span>
wget https://github.com/huangnengCSU/compleasm/releases/download/v0.2.4/compleasm-0.2.4_x64-linux.tar.bz2
tar -jxvf compleasm-0.2.4_x64-linux.tar.bz2

<span class="hljs-comment"># Install pandas if necessary</span>
pip3 install pandas                               <span class="hljs-comment"># or conda install pandas</span>

<span class="hljs-comment"># Run compleasm if lineage is known</span>
compleasm_kit/compleasm.py download primates      <span class="hljs-comment"># download data to mb_download/</span>
compleasm_kit/compleasm.py run -t16 -l primates -a hg38.fa -o hg38-mb  <span class="hljs-comment"># run the pipeline</span>

<span class="hljs-comment"># Automatically detect lineage (requiring sepp)</span>
conda install -c bioconda sepp                    <span class="hljs-comment"># if sepp hasn't been installed</span>
compleasm_kit/compleasm.py run --autolineage -a hg38.fa -o hs38-mb  
</div></code></pre>
<h2 id="contents">Contents</h2>
<ul>
<li><a href="#getting-started">Getting Started</a></li>
<li><a href="#installation">Installation</a>
<ul>
<li><a href="#conda-installation">Conda Installation</a></li>
<li><a href="#docker-installation">Docker Installation</a></li>
<li><a href="#singularity-installation">Singularity Installation</a></li>
<li><a href="#release-installation">Release Installation</a></li>
<li><a href="#manual-installation">Manual Installation</a></li>
</ul>
</li>
<li><a href="#running">Running</a>
<ul>
<li><a href="#main-modules">Main Modules</a></li>
<li><a href="#using-run-submodule-to-evaluate-genome-completeness-from-genome-assembly">Using <code>run</code> submodule to evaluate genome completeness from genome assembly</a></li>
<li><a href="#using-analyze-submodule-to-evaluate-genome-completeness-from-provided-miniprot-alignment">Using <code>analyze</code> submodule to evaluate genome completeness from provided miniprot alignment</a></li>
<li><a href="#using-download-submodule-to-download-lineage">Using <code>download</code> submodule to download lineage</a></li>
<li><a href="#using-miniprot-submodule-to-run-miniprot-alignment">Using <code>miniprot</code> submodule to run miniprot alignment</a></li>
<li><a href="#using-list-submodule-to-show-local-or-remote-busco-lineages">Using <code>list</code> submodule to show local or remote Busco lineages</a></li>
<li><a href="#using-protein-submodule-to-assess-the-completeness-of-input-proteins">Using <code>protein</code> submodule to assess the completeness of input proteins</a></li>
<li><a href="#output-description">Output description</a></li>
</ul>
</li>
<li><a href="#cite-compleasm">Citing compleasm</a></li>
</ul>
<h2 id="installation">Installation</h2>
<p>Compleasm is developed on python3.</p>
<ul>
<li>Prequisites:<br>
<a href="https://www.python.org">python3.*</a><br>
<a href="https://github.com/lh3/miniprot">miniprot</a><br>
<a href="http://hmmer.org/">hmmer</a><br>
<a href="https://github.com/smirarab/sepp">sepp</a></li>
<li>Dependencies:<br>
<a href="https://pandas.pydata.org/docs/getting_started/install.html#installing-from-pypi">pandas</a></li>
</ul>
<h3 id="conda-installation">Conda Installation</h3>
<p>Compleasm can be installed with conda. If you don't have conda, please install <a href="https://docs.conda.io/en/latest/miniconda.html">miniconda</a> or <a href="https://www.anaconda.com/products/individual">anaconda</a> first. Then you can create a new environment with compleasm installed.</p>
<pre class="hljs"><code><div>conda create -n &lt;your_env_name&gt; -c conda-forge -c bioconda compleasm
conda activate &lt;your_env_name&gt;
compleasm -h
</div></code></pre>
<h3 id="docker-installation">Docker Installation</h3>
<p>Compleasm can be installed with docker. If you don't have docker, please install <a href="https://docs.docker.com/get-docker/">docker</a> first. Then you can pull the docker image with compleasm installed.</p>
<pre class="hljs"><code><div>VERSION=0.2.2
docker run huangnengcsu/compleasm:v${VERSION} compleasm -h
</div></code></pre>
<h3 id="singularity-installation">Singularity Installation</h3>
<p>Compleasm can be installed with singularity. If you don't have singularity, please install <a href="https://docs.sylabs.io/guides/3.9/user-guide/quick_start.html#quick-installation-steps">singularity</a> first. Then you can pull the singularity image with compleasm installed.</p>
<pre class="hljs"><code><div>VERSION=0.2.2
singularity exec docker://huangnengcsu/compleasm:v${VERSION} compleasm -h
</div></code></pre>
<h3 id="release-installation">Release Installation</h3>
<pre class="hljs"><code><div>wget https://github.com/huangnengCSU/compleasm/releases/download/v0.2.4/compleasm-0.2.4_x64-linux.tar.bz2
tar -jxvf compleasm-0.2.4_x64-linux.tar.bz2
compleasm_kit/compleasm.py -h
</div></code></pre>
<h3 id="manual-installation">Manual Installation</h3>
<h4 id="get-compleasm">Get compleasm:</h4>
<pre class="hljs"><code><div>git clone https://github.com/huangnengCSU/compleasm.git
</div></code></pre>
<p>You can run the <code>compleasm.py</code> script directly or copy it to other locations then run it.</p>
<h4 id="install-miniprot">Install miniprot:</h4>
<pre class="hljs"><code><div>git clone https://github.com/lh3/miniprot
cd miniprot &amp;&amp; make
</div></code></pre>
<h4 id="install-hmmer">Install hmmer:</h4>
<pre class="hljs"><code><div>wget http://eddylab.org/software/hmmer/hmmer.tar.gz 
tar zxf hmmer.tar.gz
cd hmmer-3.3.2
./configure --prefix /your/install/path
make
make check
make install
</div></code></pre>
<h4 id="install-sepp">Install sepp:</h4>
<pre class="hljs"><code><div>git clone https://github.com/smirarab/sepp.git
cd sepp
python setup.py config -c
python setup.py install
</div></code></pre>
<h2 id="running">Running</h2>
<h3 id="main-modules">Main Modules:</h3>
<pre class="hljs"><code><div>run             Run compleasm including miniprot alignment and completeness evaluation
analyze         Evaluate genome completeness from provided miniprot alignment
download        Download specified BUSCO lineage
list            List local or remote BUSCO lineages
miniprot        Run miniprot alignment
</div></code></pre>
<h3 id="using-run-submodule-to-evaluate-genome-completeness-from-genome-assembly">Using <code>run</code> submodule to evaluate genome completeness from genome assembly:</h3>
<p>This will download the specified lineage (or automatically search for the best lineage with autolineage mode), align the
protein sequences in the lineage file to the genome sequence with miniprot, and parse the miniprot alignment result to
evaluate genome completeness.</p>
<h4 id="usage">Usage:</h4>
<pre class="hljs"><code><div>python compleasm.py run [-h] -a ASSEMBLY_PATH -o OUTPUT_DIR [-t THREADS] 
                        [-l LINEAGE] [-L LIBRARY_PATH] [-m {lite,busco}] [--specified_contigs SPECIFIED_CONTIGS [SPECIFIED_CONTIGS ...]] 
                        [--miniprot_execute_path MINIPROT_EXECUTE_PATH] [--hmmsearch_execute_path HMMSEARCH_EXECUTE_PATH] 
                        [--autolineage] [--sepp_execute_path SEPP_EXECUTE_PATH] 
                        [--min_diff MIN_DIFF] [--min_identity MIN_IDENTITY] [--min_length_percent MIN_LENGTH_PERCENT] 
                        [--min_complete MIN_COMPLETE] [--min_rise MIN_RISE]
</div></code></pre>
<h4 id="important-parameters">Important parameters:</h4>
<pre class="hljs"><code><div>  -a, --assembly_path        Input genome file in FASTA format
  -o, --output_dir           The output folder
  -t, --threads              Number of threads to use
  -l, --lineage              Specify the name of the BUSCO lineage to be used. (e.g. eukaryota, primates, saccharomycetes etc.)
  -L, --library_path         Folder path to download lineages or already downloaded lineages. 
                             If not specified, a folder named &quot;mb_downloads&quot; will be created on the current running path by default to store the downloaded lineage files.
  -m, --mode                 The mode of evaluation. Default mode is busco. 
                             lite:  Without using hmmsearch to filtering protein alignment.
                             busco: Using hmmsearch on all candidate predicted proteins to purify the miniprot alignment to improve accuracy.
  --specified_contigs        Specify the contigs to be evaluated, e.g. chr1 chr2 chr3. If not specified, all contigs will be evaluated.
  --outs                     output if score at least FLOAT*bestScore [0.95]
  --miniprot_execute_path    Path to miniprot executable file. 
                             If not specified, compleasm will search for miniprot in the directory where compleasm.py is located, the current execution directory, and system environment variables.
  --hmmsearch_execute_path   Path to hmmsearch executable file.
                             If not specified, compleasm will search for hmmsearch in the directory where compleasm.py is located, the current execution directory, and system environment variables.
  --autolineage              Automatically search for the best matching lineage without specifying lineage file.
  --sepp_execute_path        Path to sepp executable file. This is required if you want to use the autolineage mode.
</div></code></pre>
<h4 id="threshold-parameters">Threshold parameters:</h4>
<pre class="hljs"><code><div>  --min_diff               The thresholds for the best matching and second best matching. default=0.2
  --min_identity           The identity threshold for valid mapping results. default=0.4
  --min_length_percent     The fraction of protein for valid mapping results. default=0.6
  --min_complete           The length threshold for complete gene. default=0.9
</div></code></pre>
<h4 id="example">Example:</h4>
<pre class="hljs"><code><div># with lineage specified
python compleasm.py run -a genome.fasta -o output_dir -l eukaryota -t 8

# autolineage mode
python compleasm.py run -a genome.fasta -o output_dir -t 8 --autolineage

# with custom specified already downloaded lineage folder
python compleasm.py run -a genome.fasta -o output_dir -l eukaryota -t 8 -L /path/to/lineages_folder

# specify contigs
python compleasm.py run -a genome.fasta -o output_dir -l eukaryota -t 8 --specified_contigs chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22
</div></code></pre>
<h3 id="using-analyze-submodule-to-evaluate-genome-completeness-from-provided-miniprot-alignment">Using <code>analyze</code> submodule to evaluate genome completeness from provided miniprot alignment:</h3>
<p>This will directly parse the provided miniprot alignment result to evaluate genome completeness. The execute command of miniprot should be like <code>miniprot --trans -u -I --outs=0.95 --gff -t 8 ref-file protein.faa &gt; output.gff</code>.</p>
<h4 id="usage">Usage:</h4>
<pre class="hljs"><code><div>python compleasm.py analyze [-h] -g GFF -l LINEAGE -o OUTPUT_DIR [-t THREADS] [-L LIBRARY_PATH] 
                            [-m {lite,busco}] [--hmmsearch_execute_path HMMSEARCH_EXECUTE_PATH]
                            [--specified_contigs SPECIFIED_CONTIGS [SPECIFIED_CONTIGS ...]] 
                            [--min_diff MIN_DIFF] [--min_identity MIN_IDENTITY] [--min_length_percent MIN_LENGTH_PERCENT] 
                            [--min_complete MIN_COMPLETE] [--min_rise MIN_RISE]
</div></code></pre>
<h4 id="important-parameters">Important parameters:</h4>
<pre class="hljs"><code><div>  -g, --gff                 Miniprot output gff file
  -l, --lineage             BUSCO lineage name
  -o, --output_dir          Output analysis folder
  -t, --threads             Number of threads to use
  -L, --library_path        Folder path to stored lineages.
  -m, --mode                The mode of evaluation. Default mode is fast. 
                            lite:  Without using hmmsearch to filtering protein alignment.
                            busco: Using hmmsearch on all candidate predicted proteins to purify the miniprot alignment to improve accuracy.
  --hmmsearch_execute_path  Path to hmmsearch executable
                            If not specified, compleasm will search for hmmsearch in the directory where compleasm.py is located, the current execution directory, and system environment variables.
  --specified_contigs       Specify the contigs to be evaluated, e.g. chr1 chr2 chr3. If not specified, all contigs will be evaluated.
</div></code></pre>
<p>Threshold parameters are same as <code>run</code> module.</p>
<h4 id="example">Example:</h4>
<pre class="hljs"><code><div># analysis with miniprot output gff file
python compleasm.py analyze -g miniprot.gff -o output_dir -l eukaryota -t 8

# specify contigs
compleasm analyze -g miniprot.gff -o output_dir -l eukaryota -t 8 --specified_contigs chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22
</div></code></pre>
<h3 id="using-download-submodule-to-download-lineage">Using <code>download</code> submodule to download lineage:</h3>
<p>This will download the specified lineages and save to the specified folder.</p>
<h4 id="usage">Usage:</h4>
<pre class="hljs"><code><div>python compleasm.py download [-h] [-L LIBRARY_PATH] lineages [lineages ...]
</div></code></pre>
<h4 id="parameters">Parameters:</h4>
<pre class="hljs"><code><div>positional arguments:  
  lineages                Specify the names of the BUSCO lineages to be downloaded. (e.g. eukaryota, primates, saccharomycetes etc.)

optional arguments:
  -L, --library_path      The destination folder to store the downloaded lineage files.
                          If not specified, a folder named &quot;mb_downloads&quot; will be created on the current running path by default.
</div></code></pre>
<h4 id="example">Example:</h4>
<pre class="hljs"><code><div>python compleasm.py download saccharomycetes primates brassicales -L /path/to/lineages_folder
</div></code></pre>
<p>or</p>
<pre class="hljs"><code><div>python compleasm.py download saccharomycetes,primates,brassicales -L /path/to/lineages_folder
</div></code></pre>
<h3 id="using-miniprot-submodule-to-run-miniprot-alignment">Using <code>miniprot</code> submodule to run miniprot alignment:</h3>
<p>This will run miniprot alignment and output the gff file.</p>
<h4 id="usage">Usage:</h4>
<pre class="hljs"><code><div>python compleasm.py miniprot [-h] -a ASSEMBLY -p PROTEIN -o OUTDIR [-t THREADS] [--miniprot_execute_path MINIPROT_EXECUTE_PATH]
</div></code></pre>
<h4 id="important-parameters">Important parameters:</h4>
<pre class="hljs"><code><div>  -a, --assembly             Input genome file in FASTA format
  -p, --protein              Input protein file
  -o, --outdir               Miniprot alignment output directory
  -t, --threads              Number of threads to use
  --outs                     output if score at least FLOAT*bestScore [0.95]
  --miniprot_execute_path    Path to miniprot executable file. 
                             If not specified, compleasm will search for miniprot in the directory where compleasm.py is located, the current execution directory, and system environment variables.
</div></code></pre>
<h4 id="example">Example:</h4>
<pre class="hljs"><code><div>python compleasm.py miniprot -a genome.fasta -p protein.faa -o output_dir -t 8
</div></code></pre>
<h3 id="using-list-submodule-to-show-local-or-remote-busco-lineages">Using <code>list</code> submodule to show local or remote Busco lineages:</h3>
<p>This will list the local or remote BUSCO lineages.</p>
<h4 id="usage">Usage:</h4>
<pre class="hljs"><code><div>python compleasm.py list [-h] [--remote] [--local] [-L LIBRARY_PATH]
</div></code></pre>
<h4 id="important-parameters">Important parameters:</h4>
<pre class="hljs"><code><div>  --remote             List remote BUSCO lineages
  --local              List local BUSCO lineages
  -L, --library_path   Folder path to stored lineages.
</div></code></pre>
<h4 id="example">Example</h4>
<pre class="hljs"><code><div># list local lineages
python compleasm.py list --local -L /path/to/lineages_folder

# list remote lineages
python compleasm.py list --remote
</div></code></pre>
<h3 id="using-protein-submodule-to-assess-the-completeness-of-input-proteins">Using <code>protein</code> submodule to assess the completeness of input proteins:</h3>
<p>This will evaluate the completeness of input proteins.</p>
<h4 id="usage">Usage:</h4>
<pre class="hljs"><code><div>python compleasm.py protein [-h] -p PROTEINS -l LINEAGE -o OUTDIR [-t THREADS]
                            [-L LIBRARY_PATH]
                            [--hmmsearch_execute_path HMMSEARCH_EXECUTE_PATH]
</div></code></pre>
<h4 id="important-parameters">Important parameters:</h4>
<pre class="hljs"><code><div>-p, --proteins             Input protein file
-l, --lineage              BUSCO lineage name
-o, --outdir               Output analysis folder
-t, --threads              Number of threads to use
-L, --library_path         Folder path to stored lineages
--hmmsearch_execute_path   Path to hmmsearch executable.
                           If not specified, compleasm will search for miniprot in the directory where compleasm.py is located, the current execution directory, and system environment variables.
</div></code></pre>
<h4 id="example">Example:</h4>
<pre class="hljs"><code><div>python compleasm.py protein -p input.faa -l eukaryota -t 8 -o output_dir
</div></code></pre>
<h3 id="output-description">Output description</h3>
<p>The assessment result by compleasm is saved in the file <code>summary.txt</code> in the output folder. These BUSCO genes are categorized into the following classes:</p>
<ul>
<li><strong>S</strong> (<strong>Single Copy Complete Genes</strong>): The BUSCO genes that can be entirely aligned in the assembly, with only one copy present.</li>
<li><strong>D</strong> (<strong>Duplicated Complete Genes</strong>): The BUSCO genes that can be completely aligned in the assembly, with more than one copy present.</li>
<li><strong>F</strong> (<strong>Fragmented Genes, subclass 1</strong>): The BUSCO genes which only a portion of the gene is present in the assembly, and the rest of the gene cannot be aligned.</li>
<li><strong>I</strong> (<strong>Fragmented Genes, subclass 2</strong>): The BUSCO genes in which a section of the gene aligns to one position in the assembly, while the remaining part aligns to another position.</li>
<li><strong>M</strong> (<strong>Missing Genes</strong>): The BUSCO genes with no alignment present in the assembly.</li>
</ul>
<h2 id="cite-compleasm">Cite compleasm</h2>
<p>If you use compleasm, please cite:</p>
<blockquote>
<p>Neng Huang, Heng Li, compleasm: a faster and more accurate reimplementation of BUSCO. <em>Bioinformatics</em>, <strong>39</strong>, btad595, 2023. <a href="https://doi.org/10.1093/bioinformatics/btad595">doi:10.1093/bioinformatics/btad595</a></p>
</blockquote>

</body>
</html>
